# FEATURE - validate message body

blackironj opened the pull request:
- 각 메시지 항목의 타입 및 길이를 확인하도록 함.
- General service에서 메시지를 parsing 후 메시지의 유효성 검사까지 하도록 수정.

Commit:
# FEATURE - validate message body

- 각 메시지 항목의 type 및 길이를 확인하도록 함.

morishjs commented on src/plugins/net_plugin/rpc_services/rpc_services.cpp:
```diff
@@ -171,6 +181,113 @@ bool GeneralService::validateMsgFormat(MessageHeader *header) {
+    return false;
+  }
+  case MsgEntryType::STRING: {
+    if (msg_body.find(key) != msg_body.end() && msg_body[key].is_string())
```
msg_body.contains(key)로 간단하게 포함여부를 알 수 있습니다.

blackironj replied:
수정하였습니다.

morishjs commented on src/plugins/net_plugin/rpc_services/rpc_services.cpp:
```diff
@@ -171,6 +181,113 @@ bool GeneralService::validateMsgFormat(MessageHeader *header) {
+
+  if (MSG_VALID_FILTER.count(msg_type) > 0) {
+    for (auto &filt_entry : MSG_VALID_FILTER.at(msg_type)) {
+      if (!isValidMsgEntryType(json_body, get<0>(filt_entry), get<1>(filt_entry)) ||
```
auto &[key, value] : MSG_VALID_FILTER.at(msg_type) 가 더 가독성이 좋아보입니다.

blackironj replied:
auto &[key, type, len] : MSG_VALID_FILTER.at(msg_type) 형태로 수정하였습니다.

morishjs commented on src/plugins/net_plugin/config/include/message.hpp:
```diff
@@ -88,5 +90,34 @@ struct MessageHeader {
+
+using EntryAttribute = std::tuple<std::string, MsgEntryType, MsgEntryLength>;
+
+const std::map<MessageType, std::vector<EntryAttribute>> MSG_VALID_FILTER = {
```
unordered_map이나 map이나 검색속도에서 유의미한 성능차이는 없을 것 같은데 그래도 map을 써야하는 특별한 이유가 없다면 unordered_map가 낫지 않을까요? O(logn) vs O(1)

blackironj replied:
unordered_map으로 수정하였습니다.

Commit:
# EDIT - code readability

- msg_body.find(key) != msg_body.end() => msg_body.contains(key)
- auto &filt_entry => auto &[key, type, len]
- MSG_VALID_FILTER ( map => unordered_map )
